### ISSUE TITLE
Advanced protection sporadicaly stops working entirely

### ISSUE THREAD

**Issue by Morku** [AdguardTeam/AdguardForiOS#2204]
- [X] I am running the latest version
- [X] I checked the documentation and found no answer
- [X] I checked to make sure that this issue has not already been filed

### Issue Details

- **AdGuard version:**
  - 4.5.0.1002
- **Device model and storage size:**
  - iPhone XS Max 64GB
- **Operating system and version:**
  - 16.6
- **Browser or App:**
  - Safari

The Advanced Protection sporadicly stops working for unknown reason. The Safari content filters are unaffected. They continue working properly.
There is no way to reproduce it. It simply stops at some time. Sometimes after 2-3 weeks, sometimes 1-2 times within a week. It might have to do with/after network changes or filtering update.
THats why I struggle over month to fill the Bug report, because it can take weeks/month to report back if try something.

When the issue starts to happen, the App still act like that Advanced Protection is still enabled and even the advanced option shows the amount of active filters.

**To solve the issue I have tried:**

- update Safari filters -> unsolved
- Dis-/Enable Safari filters and waited to be updated -> unsolved
- Dis-/Enable Advanced Protection only and waited to update -> unsolved
- Dis-/Enable AdGuard Extension in Safarai -> unsolved
- Dis-/Enable content blockers in Safarai -> unsolved
- Closed all Tabs in Safarai and closed Safari itself -> unsolved
- Clear Cookies/Cache in Safari (however, I always browse and use Private mode without traces) -> unsolved
- flight mode On/Off -> unsolved
- changing network -> unsolved
- Turn Off/On iPhone -> unsolved
- Reset AdGuard Settings -> unsolved

**And this is the only solution I found to fix the issue:**

- Uninstall AdGuard completely and redownload from App Store -> configure from beginning -> solved

until it occurs again after Days/Weeks/Months

### Video
Here are two example pages where Advanced protection usually works fine:
https://www.sueddeutsche.de/politik/ukraine-binnenfluechtlinge-reportage-krieg-russland-1.6126900?reduced=true

https://www.videohelp.com/software?updated=List+all+by+latest+version

Both filters are CSS extended filters `#?#` that are handled by Advanced Protection. On Sueddeutsche there is that "-Anazeige-" Ad placeholder. On videohelp there is that "Try StreamFab" Adline. Both shouldn't be displayed with AdGuard filters.

<details><summary>Video:</summary>

https://github.com/AdguardTeam/AdguardForiOS/assets/6276915/cc3983b9-7ed1-4d1c-baa1-c32a11891d98

</details>

### Additional Information

I have made a Debug Logfile ater I got back the issue and browsed both pages. Tell me, if needed.

**Comment by Versty** [comment]
@Morku Thank you for reporting! 
Have you noticed this problem in versions prior to 4.5?

The logs from you device would be much appreciated. Please send it to apple@adguard.com and mention the number `2204` in the subject line.

**Comment by Morku** [comment]
@Versty 
I have send you the logfile at the moment, the issue occur.

Yes, I noticed the issue already prior 4.5, but was never confident enough where and when the issue happen and did not open a ticket in hope that "it doesn't happen again".
At the moment, everything is back fine, after the Advanced protection recovered again by waiting 2 Days.

**Comment by Morku** [comment]
Today the issue recovered. Yes, it took that long (about 1 month). Again for no obvious reason. No filter of Advanced protection is working (most likely noticable with several Cookie Pop-ups), but content filters works fine.
Updating filters, restart Safari does not fix the issue (see list above).
AdGuard is not showing any error or reason why it shouldn't work (Advanced protection is showing 22046 active rules).
I have sent another logfile.

**Comment by Versty** [comment]
@Morku Could you recall the time when you noticed that the Advanced rules were not working?

**Comment by Morku** [comment]
@Versty 
About ~30 minutes or more before I created the second logfile. So maybe 14th September 11:00am (+- 30 minutes).
When I noticed I enabled Debug Logs again and browsed 2-3 more pages to check if there is still the issue. Then I created the logfile.

On the next day, 15th September, everything started working fine again.

**Comment by Versty** [comment]
@Morku Has [this](https://github.com/AdguardTeam/AdguardForiOS/issues/2188#issuecomment-1631249736)  error was present prior to the problem occurring?

**Comment by Morku** [comment]
@Versty 
No, I have never seen this error before. Not in iOS 16 and not in iOS 17.

I sometimes have the "updating filters failed" error, but it's unrelated to the issue, that Advanced protection stops working.

**Comment by Versty** [comment]
@Morku Thank you for the provided information. We are investigating.

**Comment by Morku** [comment]
@Versty 

Hi, I want to tell that the issue still persist with iOS 17.2 and AdGuard for iOS 4.5.2.

It looks like the issue happens when:

- I change network connection to other WiFi or mobile cellular. I use AdGuard DNS for lokal VPN filtering, except for my own WiFi (maybe it's related).
- Or when I test around with filters and dis-/enable filterlists to test things. When I all enable back and make sure in settings that "all filters are active".

However, one thing seems improved since iOS 17.0 -> not all Advanced Protection rules are affected anymore. Cookie Pop-ups (scriptlet) do work always fine now. Might be luck(?)

When the issue recover, I notice for two pages:

https://www.videohelp.com/software?updated=List+all+by+latest+version

<details><summary>Screenshot-1:</summary>

![IMG_724699475271](https://github.com/AdguardTeam/AdguardForiOS/assets/6276915/42db6ac3-1c3d-4256-a872-18ef33f0df7a)


</details><br/>

This text Ad is handled by:
`videohelp.com#?#div[id] a[href]:contains(/^Try (?:D.?V.?D.?F.?a.?b.?|StreamFab)|^StreamFab|^DVDFab$/) + a[href]:upward(1)`

https://www.sueddeutsche.de/wirtschaft/bahn-streik-gdl-januar-1.6321748?reduced=true

<details><summary>Screenshot-2:</summary>

![IMG_724699648624](https://github.com/AdguardTeam/AdguardForiOS/assets/6276915/2dbcb677-af64-4cae-9d28-cff34d21b51a)


</details><br/>

This placeholder is handled by:
`sueddeutsche.de#?#article > div[data-manual-remove="true"]:has(> div[id^="iqadt"])`

It is noticable that rules with `#?#` are affected.

It is still uncontrollable when my phone decide to recover the Extension. Sometimes hours, sometimes a Day ... It is odd. I hope the Info helps.

**Comment by Versty** [comment]
@Morku Thank you for additional details. 
This issue will be fixed with a complete code rework in one of the next major updates.

**Comment by Krisnapong** [comment]
Happened to me on both iPads, but not iPhone. At the same time. Twice, on both of them. And just happened again tonight, on both of them, and again, not on iPhone. That’s why I’m here. 

Hate to re-install and redo all the setup again though. On BOTH iPads. 

Truly hope that this problem will be fixed soon. It’s already year 2025.

**Comment by Krisnapong** [comment]
Happened to me twice on both iPads, but not iPhone. And both were having this issue at the same time. 

Truly hope that this problem will be fixed soon. It’s already year 2025.

**Comment by Krisnapong** [comment]
By the way, if this issue is in the medium priority to get fixed, is it possible to at least implement the “save setting” function? At least that would help with the re-installing the app, which is the only solution to this problem. The problem still persists and it is such a pain to re-install the app and redo all the setting. 

And out of curiosity, why the bug that renders the important function of the app useless is considered a medium priority? Advance protection is pretty much the function that makes this app very effective.

**Comment by Krisnapong** [comment]
This issue is just happened again on one of my iPad. The other iPad is fine. The iPhone is fine. All devices have just been updated to AdGuard for iOS 4.5.10 yesterday. 

Seems to get this bug randomly without any suspicious activity. Normal web browsing, not installing new app, didn’t touch on any configuration or settings of AdGuard nor iOS. It just suddenly broke! Again! 

It’s really annoying. AdGuard has no “backup settings” functionality on iOS app. All filters, user rules, allow lists, DNS filters, bootstrap server, fallback server have to be redone by hand from scratch after reinstalling the app … every single time.

**Comment by ihurin** [comment]
ZD tickets 1099451 & 1099427 reported the same problem.

**Comment by Morku** [comment]
Since SafariConverterLib got a major improvement, I haven't seen that bug anymore.
It might got resolved.
